README.md
# CDK Construct Website

An AWS CDK construct for hosting websites on AWS S3

- Creates a private bucket for hosting static assets
- Assert no buckets are configured as websites
- Creates an origin access identity (OAI)
- Assert bucket grants read access to origin
- Creates a distribution
- Configures distribution origin access identity (OAI)
- Configures distribution custom error responses
- Allows domain name configuration
- Enables distribution logging 
- Default log expiration to 14 days
- Allow log life cycle configuration

## Install

`npm install --save cdk-construct-website`


## Example Using AWS S3 Deployment 

```typescript
import * as cdk from '@aws-cdk/core';
import { Website } from 'cdk-construct-website';
import { BucketDeployment, Source } from '@aws-cdk/aws-s3-deployment';

export class ExampleStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);
    const website = new Website(this, 'Website');

    new BucketDeployment(this, 'BucketDeployment', {
      sources: [Source.asset('./dist')],
      destinationBucket: website.bucket,
      distribution: website.distribution,
      distributionPaths: ['/*'],
    });
  }
}
```

## Full Example

```typescript
import * as cdk from '@aws-cdk/core';
import { Website } from 'cdk-construct-website';
import { BucketDeployment, Source } from '@aws-cdk/aws-s3-deployment';
import { HostedZone, ARecord, RecordTarget } from '@aws-cdk/aws-route53';
import { CloudFrontTarget } from '@aws-cdk/aws-route53-targets';
import { DnsValidatedCertificate } from '@aws-cdk/aws-certificatemanager';

export class ExampleStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const domainName = new cdk.CfnParameter(this, 'DomainName', {
      type: 'String',
      description: 'Domain name',
    })

    const hostedZoneId = new cdk.CfnParameter(this, 'HostedZoneId', {
      type: 'String',
      description: 'Route53 Hosted Zone Id'
    })

    const hostedZone = HostedZone.fromHostedZoneAttributes(this, 'HostedZone', {
      hostedZoneId: hostedZoneId.valueAsString,
      zoneName: domainName.valueAsString
    })

    const cert = new DnsValidatedCertificate(this, 'AwsManagedCertificate', {
      domainName: domainName.valueAsString,
      hostedZone: hostedZone,
      subjectAlternativeNames: [`*.${domainName.valueAsString}`]
    });

    const website = new Website(this, 'Website', {
      domainName: domainName.valueAsString,
      certificateArn: cert.certificateArn
    });

    new BucketDeployment(this, 'WebDeployment', {
      sources: [Source.asset('./dist')],
      destinationBucket: website.bucket,
      distribution: website.distribution,
      distributionPaths: ['/*'],
    });

    new ARecord(this, 'WebAliasRecord', {
      zone: hostedZone,
      recordName: domainName.valueAsString,
      target: RecordTarget.fromAlias(new CloudFrontTarget(website.distribution)),
    });
  }
}
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.7](https://github.com/hollanddd/cdk-construct-website/compare/v0.0.6...v0.0.7) (2021-04-03)

### [0.0.6](https://github.com/hollanddd/cdk-construct-website/compare/v0.0.5...v0.0.6) (2021-03-25)
### 0.0.6 (2021-04-02)


### Features

* website hosting ([04a8a8e](https://github.com/hollanddd/cdk-construct-website/commit/04a8a8e29a1a190466ec1157a6b6e8c703cf08f7))

### Bug Fixes

* anti tamper warning ([1860243](https://github.com/hollanddd/cdk-construct-website/commit/186024335db2fe1d8bf5cfe7c9b0a74cafb1494b))
* **docs:** provide working example ([6bd481e](https://github.com/hollanddd/cdk-construct-website/commit/6bd481eb5bbf07e95896b43d6d81a53cdaadfc90))
* compile errors ([831b940](https://github.com/hollanddd/cdk-construct-website/commit/831b94005bfa2f6d0bdb8bd6ae09b2606723c437))

### [0.0.5](https://github.com/hollanddd/cdk-construct-website/compare/v0.0.4...v0.0.5) (2021-03-18)

### [0.0.4](https://github.com/hollanddd/cdk-construct-website/compare/v0.0.3...v0.0.4) (2021-03-06)

### [0.0.3](https://github.com/hollanddd/cdk-construct-website/compare/v0.0.2...v0.0.3) (2021-03-06)


### Bug Fixes

* **docs:** provide working example ([6bd481e](https://github.com/hollanddd/cdk-construct-website/commit/6bd481eb5bbf07e95896b43d6d81a53cdaadfc90))

### [0.0.2](https://github.com/hollanddd/cdk-construct-website/compare/v0.0.1...v0.0.2) (2021-03-06)

### 0.0.1 (2021-03-06)


### Features

* website hosting ([04a8a8e](https://github.com/hollanddd/cdk-construct-website/commit/04a8a8e29a1a190466ec1157a6b6e8c703cf08f7))


### Bug Fixes

* compile errors ([831b940](https://github.com/hollanddd/cdk-construct-website/commit/831b94005bfa2f6d0bdb8bd6ae09b2606723c437))


test/website.test.ts
import '@aws-cdk/assert/jest';
import { App, Stack, Duration } from '@aws-cdk/core';
import { Website } from '../src/index';
import {
  getBucketFromStack,
  getOriginAccessIdentityFromStack,
} from './helpers/cdk.helpers';

test('Creates a private bucket for hosting static assets', () => {
  const mockApp = new App();
  const stack = new Stack(mockApp, 'testing-stack');

  new Website(stack, 'testing');

  expect(stack).toHaveResource('AWS::S3::Bucket', {
    PublicAccessBlockConfiguration: {
      BlockPublicAcls: true,
      BlockPublicPolicy: true,
      IgnorePublicAcls: true,
      RestrictPublicBuckets: true,
    },
  });
});

test('Assert no buckets are configured as websites', () => {
  const mockApp = new App();
  const stack = new Stack(mockApp, 'testing-stack');

  new Website(stack, 'testing');

  expect(stack).not.toHaveResource('AWS::S3::Bucket', {
    WebsiteConfiguration: {},
  });
});

test('Creates an origin access identity (OAI)', () => {
  const mockApp = new App();
  const stack = new Stack(mockApp, 'testing-stack');

  new Website(stack, 'testing');

  expect(stack).toHaveResource('AWS::CloudFront::CloudFrontOriginAccessIdentity');
});

test('Assert bucket grants read access to origin', () => {
  const mockApp = new App();
  const stack = new Stack(mockApp, 'testing-stack');

  new Website(stack, 'testing');

  const bucket = getBucketFromStack(stack, 'testing-assets-bucket');
  const bucketId = stack.getLogicalId(bucket);
  const origin = getOriginAccessIdentityFromStack(stack, 'testing-oai');

  expect(stack).toHaveResourceLike('AWS::S3::BucketPolicy', {
    Bucket: {
      Ref: bucketId,
    },
    PolicyDocument: {
      Statement: [
        {
          Action: [
            's3:GetObject*',
            's3:GetBucket*',
            's3:List*',
          ],
          Effect: 'Allow',
          Principal: {
            CanonicalUser: {
              'Fn::GetAtt': [
                stack.getLogicalId(origin),
                'S3CanonicalUserId',
              ],
            },
          },
          Resource: [
            {
              'Fn::GetAtt': [
                bucketId,
                'Arn',
              ],
            },
            {
              'Fn::Join': [
                '',
                [
                  {
                    'Fn::GetAtt': [
                      bucketId,
                      'Arn',
                    ],
                  },
                  '/*',
                ],
              ],
            },
          ],
        },
      ],
    },
  });
});

test('Creates a distribution', () => {
  const mockApp = new App();
  const stack = new Stack(mockApp, 'testing-stack');

  new Website(stack, 'testing');

  expect(stack).toHaveResourceLike('AWS::CloudFront::Distribution', {
    DistributionConfig: {
      Enabled: true,
    },
  });
});

test('Configures distribution origin access identity (OAI)', () => {
  const mockApp = new App();
  const stack = new Stack(mockApp, 'testing-stack');

  new Website(stack, 'testing');

  let origin = getOriginAccessIdentityFromStack(stack, 'testing-oai');

  expect(stack).toHaveResourceLike('AWS::CloudFront::Distribution', {
    DistributionConfig: {
      Origins: [{
        S3OriginConfig: {
          OriginAccessIdentity: {
            'Fn::Join': [
              '',
              [
                'origin-access-identity/cloudfront/',
                { Ref: stack.getLogicalId(origin) },
              ],
            ],
          },
        },
      }],
    },
  });
});

test('Configures distribution custom error responses', () => {
  const mockApp = new App();
  const stack = new Stack(mockApp, 'testing-stack');

  new Website(stack, 'testing');

  expect(stack).toHaveResourceLike('AWS::CloudFront::Distribution', {
    DistributionConfig: {
      CustomErrorResponses: [
        {
          ErrorCode: 404,
          ResponseCode: 200,
          ResponsePagePath: '/index.html',
        },
      ],
    },
  });
});
test('Allows domain name configuration', () => {
  const mockApp = new App();
  const stack = new Stack(mockApp, 'testing-stack');

  new Website(stack, 'testing', { domainName: 'example.com', certificateArn: 'some-arn' });
  expect(stack).toHaveResourceLike('AWS::CloudFront::Distribution', {
    DistributionConfig: {
      Aliases: ['example.com'],
      ViewerCertificate: {
        AcmCertificateArn: 'some-arn',
        SslSupportMethod: 'sni-only',
      },
    },
  });
});

test('Enables distribution logging', () => {
  const mockApp = new App();
  const stack = new Stack(mockApp, 'testing-stack');

  new Website(stack, 'testing');
  const bucket = getBucketFromStack(stack, 'testing-logging-bucket');

  expect(stack).toHaveResourceLike('AWS::CloudFront::Distribution', {
    DistributionConfig: {
      Logging: {
        Bucket: {
          'Fn::GetAtt': [
            stack.getLogicalId(bucket),
            'RegionalDomainName',
          ],
        },
        IncludeCookies: false,
        Prefix: 'website',
      },
    },
  });
});

test('Default log expiration to 14 days', () => {
  const mockApp = new App();
  const stack = new Stack(mockApp, 'testing-stack');

  new Website(stack, 'testing');
  expect(stack).toHaveResourceLike('AWS::S3::Bucket', {
    LifecycleConfiguration: {
      Rules: [{ ExpirationInDays: 14, Status: 'Enabled' }],
    },
  });
});

test('Allow log life cycle configuration', () => {
  const mockApp = new App();
  const stack = new Stack(mockApp, 'testing-stack');

  new Website(stack, 'testing', { logExpiration: Duration.days(31) });
  expect(stack).toHaveResourceLike('AWS::S3::Bucket', {
    LifecycleConfiguration: {
      Rules: [{ ExpirationInDays: 31, Status: 'Enabled' }],
    },
  });
});


test/helpers/cdk.helpers.ts
import { CfnCloudFrontOriginAccessIdentity } from '@aws-cdk/aws-cloudfront';
import { CfnBucket, CfnBucketPolicy } from '@aws-cdk/aws-s3';
import { Stack, ConstructNode } from '@aws-cdk/core';

// castNode is a generic for casting construct nodes to a type
function castNode<T>(node: ConstructNode): T {
  return node.defaultChild as unknown as T;
}

// stackNodeByID iterates through all children of a stack to return the first
// child node with a matching id.
function stackNodeByID(stack: Stack, id: string): ConstructNode {
  let children = stack.node.findAll();
  for (let child of children) {
    if (child.node.id === id) {
      return child.node;
    }
  }
  return stack.node;
}

export function listStackChildren(stack: Stack) {
  let children = stack.node.findAll();
  for (let child of children) {
    console.log(child.node.id);
  }
}

// getOriginAccessIdentityFromStack returns a stacks origin access identity as a
// cloud formation construct. This is used in tests to validate references to
// realated resources.
export function getOriginAccessIdentityFromStack(stack: Stack, id: string): CfnCloudFrontOriginAccessIdentity {
  return castNode<CfnCloudFrontOriginAccessIdentity>(stackNodeByID(stack, id));
}

export function getBucketFromStack(stack: Stack, id: string): CfnBucket {
  return castNode<CfnBucket>(stackNodeByID(stack, id));
}

export function getBucketPolicyFromStack(stack: Stack, id: string): CfnBucketPolicy {
  return castNode<CfnBucketPolicy>(stackNodeByID(stack, id));
}


.projenrc.js
const { AwsCdkConstructLibrary } = require('projen');

const project = new AwsCdkConstructLibrary({
  author: 'Darren',
  authorAddress: 'me@darrenholland.com',
  cdkVersion: '1.96.0',
  defaultReleaseBranch: 'main',
  jsiiFqn: 'projen.AwsCdkConstructLibrary',
  name: 'cdk-construct-website',
  repositoryUrl: 'https://github.com/hollanddd/cdk-construct-website.git',

  /* AwsCdkConstructLibraryOptions */
  // cdkAssert: true,                                                          /* Install the @aws-cdk/assert library? */
  cdkDependencies: [
    '@aws-cdk/aws-cloudfront',
    '@aws-cdk/aws-s3',
    '@aws-cdk/core',
  ],
  // cdkDependenciesAsDeps: true,                                              /* If this is enabled (default), all modules declared in `cdkDependencies` will be also added as normal `dependencies` (as well as `peerDependencies`). */
  // cdkTestDependencies: undefined,                                           /* AWS CDK modules required for testing. */
  // cdkVersionPinning: false,                                                 /* Use pinned version instead of caret version for CDK. */

  /* ConstructLibraryOptions */
  // catalog: undefined,                                                       /* Libraries will be picked up by the construct catalog when they are published to npm as jsii modules and will be published under:. */

  /* JsiiProjectOptions */
  // compat: false,                                                            /* Automatically run API compatibility test against the latest version published to npm after compilation. */
  // compatIgnore: '.compatignore',                                            /* Name of the ignore file for API compatibility tests. */
  // docgen: true,                                                             /* Automatically generate API.md from jsii. */
  // eslint: true,                                                             /* Install eslint. */
  // eslintOptions: undefined,                                                 /* Eslint options. */
  // excludeTypescript: undefined,                                             /* Accepts a list of glob patterns. */
  // publishToGo: undefined,                                                   /* Publish Go bindings to a git repository. */
  // publishToMaven: undefined,                                                /* Publish to maven. */
  // publishToNuget: undefined,                                                /* Publish to NuGet. */
  // publishToPypi: undefined,                                                 /* Publish to pypi. */
  // rootdir: '.',                                                             /* undefined */
  // sampleCode: true,                                                         /* Generate one-time sample in `src/` and `test/` if there are no files there. */

  /* NodePackageOptions */
  // allowLibraryDependencies: true,                                           /* Allow the project to include `peerDependencies` and `bundledDependencies`. */
  // authorEmail: undefined,                                                   /* Author's e-mail. */
  // authorName: undefined,                                                    /* Author's name. */
  // authorOrganization: undefined,                                            /* Author's Organization. */
  // authorUrl: undefined,                                                     /* Author's URL / Website. */
  // autoDetectBin: true,                                                      /* Automatically add all executables under the `bin` directory to your `package.json` file under the `bin` section. */
  // bin: undefined,                                                           /* Binary programs vended with your module. */
  // bundledDeps: undefined,                                                   /* List of dependencies to bundle into this module. */
  // deps: [],                                                                 /* Runtime dependencies of this module. */
  // description: undefined,                                                   /* The description is just a string that helps people understand the purpose of the package. */
  // devDeps: [],                                                              /* Build dependencies for this module. */
  // entrypoint: 'lib/index.js',                                               /* Module entrypoint (`main` in `package.json`). */
  // homepage: undefined,                                                      /* Package's Homepage / Website. */
  // keywords: undefined,                                                      /* Keywords to include in `package.json`. */
  // license: 'Apache-2.0',                                                    /* License's SPDX identifier. */
  // licensed: true,                                                           /* Indicates if a license should be added. */
  // maxNodeVersion: undefined,                                                /* Minimum node.js version to require via `engines` (inclusive). */
  // minNodeVersion: undefined,                                                /* Minimum Node.js version to require via package.json `engines` (inclusive). */
  // npmAccess: undefined,                                                     /* Access level of the npm package. */
  // npmDistTag: 'latest',                                                     /* Tags can be used to provide an alias instead of version numbers. */
  // npmRegistryUrl: 'https://registry.npmjs.org',                             /* The base URL of the npm package registry. */
  // npmTaskExecution: NpmTaskExecution.PROJEN,                                /* Determines how tasks are executed when invoked as npm scripts (yarn/npm run xyz). */
  // packageManager: NodePackageManager.YARN,                                  /* The Node Package Manager used to execute scripts. */
  // packageName: undefined,                                                   /* The "name" in package.json. */
  // peerDependencyOptions: undefined,                                         /* Options for `peerDeps`. */
  // peerDeps: [],                                                             /* Peer dependencies for this module. */
  // projenCommand: 'npx projen',                                              /* The shell command to use in order to run the projen CLI. */
  // repository: undefined,                                                    /* The repository is the location where the actual code for your package lives. */
  // repositoryDirectory: undefined,                                           /* If the package.json for your package is not in the root directory (for example if it is part of a monorepo), you can specify the directory in which it lives. */
  // scripts: {},                                                              /* npm scripts to include. */
  // stability: undefined,                                                     /* Package's Stability. */

  /* NodeProjectOptions */
  // antitamper: true,                                                         /* Checks that after build there are no modified files on git. */
  // artifactsDirectory: 'dist',                                               /* A directory which will contain artifacts to be published to npm. */
  // buildWorkflow: undefined,                                                 /* Define a GitHub workflow for building PRs. */
  // codeCov: false,                                                           /* Define a GitHub workflow step for sending code coverage metrics to https://codecov.io/ Uses codecov/codecov-action@v1 A secret is required for private repos. Configured with @codeCovTokenSecret. */
  // codeCovTokenSecret: undefined,                                            /* Define the secret name for a specified https://codecov.io/ token A secret is required to send coverage for private repositories. */
  // copyrightOwner: undefined,                                                /* License copyright owner. */
  // copyrightPeriod: undefined,                                               /* The copyright years to put in the LICENSE file. */
  // dependabot: true,                                                         /* Include dependabot configuration. */
  // dependabotOptions: undefined,                                             /* Options for dependabot. */
  // gitignore: undefined,                                                     /* Additional entries to .gitignore. */

  jest: true, /* Setup jest unit tests. */
  jestOptions: {
    ignorePatterns: ['/test/helpers'],
  }, /* Jest options. */
  //
  // jsiiReleaseVersion: 'latest',                                             /* Version requirement of `jsii-release` which is used to publish modules to npm. */
  // mergify: true,                                                            /* Adds mergify configuration. */
  // mergifyAutoMergeLabel: 'auto-merge',                                      /* Automatically merge PRs that build successfully and have this label. */
  // mergifyOptions: undefined,                                                /* Options for mergify. */
  // npmignore: undefined,                                                     /* Additional entries to .npmignore. */
  // npmignoreEnabled: true,                                                   /* Defines an .npmignore file. Normally this is only needed for libraries that are packaged as tarballs. */
  // projenDevDependency: true,                                                /* Indicates of "projen" should be installed as a devDependency. */
  // projenUpgradeAutoMerge: undefined,                                        /* Automatically merge projen upgrade PRs when build passes. */
  // projenUpgradeSchedule: [ '0 6 * * *' ],                                   /* Customize the projenUpgrade schedule in cron expression. */
  // projenUpgradeSecret: undefined,                                           /* Periodically submits a pull request for projen upgrades (executes `yarn projen:upgrade`). */
  // projenVersion: Semver.latest(),                                           /* Version of projen to install. */
  // pullRequestTemplate: true,                                                /* Include a GitHub pull request template. */
  // pullRequestTemplateContents: undefined,                                   /* The contents of the pull request template. */
  // rebuildBot: undefined,                                                    /* Installs a GitHub workflow which is triggered when the comment "@projen rebuild" is added to a pull request. */
  // rebuildBotCommand: 'rebuild',                                             /* The pull request bot command to use in order to trigger a rebuild and commit of the contents of the branch. */
  // releaseBranches: [ 'main' ],                                              /* Branches which trigger a release. */
  // releaseEveryCommit: true,                                                 /* Automatically release new versions every commit to one of branches in `releaseBranches`. */
  // releaseSchedule: undefined,                                               /* CRON schedule to trigger new releases. */
  // releaseToNpm: false,                                                      /* Automatically release to npm when new versions are introduced. */
  // releaseWorkflow: undefined,                                               /* Define a GitHub workflow for releasing from "main" when new versions are bumped. */
  // workflowBootstrapSteps: 'yarn install --frozen-lockfile && yarn projen',  /* Workflow steps to use in order to bootstrap this repo. */
  // workflowContainerImage: undefined,                                        /* Container image to use for GitHub workflows. */
  // workflowNodeVersion: undefined,                                           /* The node version to use in GitHub workflows. */

  /* ProjectOptions */
  // clobber: true,                                                            /* Add a `clobber` task which resets the repo to origin. */
  // devContainer: false,                                                      /* Add a VSCode development environment (used for GitHub Codespaces). */
  // gitpod: false,                                                            /* Add a Gitpod development environment. */
  // logging: {},                                                              /* Configure logging options such as verbosity. */
  // outdir: '.',                                                              /* The root directory of the project. */
  // parent: undefined,                                                        /* The parent project, if this project is part of a bigger project. */
  // projectType: ProjectType.UNKNOWN,                                         /* Which type of project this is (library/app). */
  // readme: undefined,                                                        /* The README setup. */
});

project.synth();


src/index.ts
import { CloudFrontWebDistribution, CloudFrontWebDistributionProps, OriginAccessIdentity } from '@aws-cdk/aws-cloudfront';
import { Bucket, BlockPublicAccess } from '@aws-cdk/aws-s3';
import { Construct, RemovalPolicy, Duration } from '@aws-cdk/core';

export interface WebsiteProps {
  readonly domainName?: string;
  readonly logExpiration?: Duration;
  readonly certificateArn?: string;
}

export class Website extends Construct {
  public distribution: CloudFrontWebDistribution
  public bucket: Bucket

  constructor(scope: Construct, id: string, props?: WebsiteProps) {
    super(scope, id);

    this.bucket = new Bucket(this, `${id}-assets-bucket`, {
      removalPolicy: RemovalPolicy.DESTROY,
      publicReadAccess: false,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
    });

    const oai = new OriginAccessIdentity(this, `${id}-oai`, {
      comment: `OAI for ${this.bucket.bucketName} website`,
    });

    this.bucket.grantRead(oai);

    let distroProps: CloudFrontWebDistributionProps = {
      errorConfigurations: [{
        errorCode: 404,
        responseCode: 200,
        responsePagePath: '/index.html',
      }],
      loggingConfig: {
        bucket: new Bucket(this, `${id}-logging-bucket`, {
          publicReadAccess: false,
          blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
          lifecycleRules: [{
            enabled: true,
            expiration: (props && props.logExpiration) ? props.logExpiration : Duration.days(14),
          }],
        }),
        prefix: 'website',
        includeCookies: false,
      },
      originConfigs: [
        {
          s3OriginSource: {
            s3BucketSource: this.bucket,
            originAccessIdentity: oai,
          },
          behaviors: [{ isDefaultBehavior: true }],
        },
      ],
    };

    if (props && props.domainName !== undefined && props.certificateArn !== undefined) {
      distroProps = this.addViewerCertificate(distroProps, [props.domainName], props.certificateArn);
    }

    this.distribution = new CloudFrontWebDistribution(this, `${id}-distribution`, distroProps);
  }

  private addViewerCertificate(props: CloudFrontWebDistributionProps, aliases: string[], acmCertificateArn: string): CloudFrontWebDistributionProps {
    return {
      viewerCertificate: {
        aliases: aliases,
        props: {
          acmCertificateArn: acmCertificateArn,
          sslSupportMethod: 'sni-only',
        },
      },
      ...props,
    };
  }
}

